File: 200.number-of-islands.cpp
#include <bits/stdc++.h>
using namespace std;
/*
 * @lc app=leetcode id=200 lang=cpp
 *
 * [200] Number of Islands
 */

// @lc code=start
class Solution
{
public:
    void visit(vector<vector<char>> &grid, int i, int j)
    {
        int n = grid.size(), m = grid[0].size();
        if (i < 0 || i == n || j < 0 || j == m || grid[i][j] != '1')
        {
            return;
        }
        grid[i][j] = '2';
        visit(grid, i + 1, j);
        visit(grid, i - 1, j);
        visit(grid, i, j + 1);
        visit(grid, i, j - 1);
    }

    int numIslands(vector<vector<char>> &grid)
    {
        int num = 0;
        for (int i = 0; i < grid.size(); i++)
        {
            for (int j = 0; j < grid[0].size(); j++)
            {
                if (grid[i][j] == '1')
                {
                    num++;
                    visit(grid, i, j);
                }
            }
        }
        return num;
    }
};
// @lc code=end

File: Java/Grind75/Week 4/200.number-of-islands.java
import java.util.*;

/*
 * @lc app=leetcode id=200 lang=java
 *
 * [200] Number of Islands
 */

// @lc code=start
class Solution {
    public int numIslands(char[][] grid) {
        boolean[][] vis = new boolean[grid.length][grid[0].length];
        int count = 0;
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                if (vis[i][j] == false && grid[i][j] == '1') {
                    count++;
                    visitIsland(vis, grid, i, j);
                }
            }
        }
        return count;
    }

    private void visitIsland(boolean[][] vis, char[][] grid, int row, int col) {
        vis[row][col] = true;
        if ((row + 1) < grid.length && grid[row + 1][col] == '1' && vis[row + 1][col] == false) {
            visitIsland(vis, grid, row + 1, col);
        }
        if ((col + 1) < grid[0].length && grid[row][col + 1] == '1' && vis[row][col + 1] == false) {
            visitIsland(vis, grid, row, col + 1);
        }
        if ((row - 1) >= 0 && grid[row - 1][col] == '1' && vis[row - 1][col] == false) {
            visitIsland(vis, grid, row - 1, col);
        }
        if ((col - 1) >= 0 && grid[row][col - 1] == '1' && vis[row][col - 1] == false) {
            visitIsland(vis, grid, row, col - 1);
        }
    }
}
// @lc code=end
